src/udev/udev-node.c

```diff
@@ -232,7 +229,42 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
                                 r = -errno;
                 } while (r == -ENOENT);
 
-        return r;
+        for (i = 0; i < LINK_UPDATE_MAX_RETRIES; i++) {
+                struct stat st1, st2 = {};
+
+                r = stat(dirname, &st1);
+                if (r < 0 && errno != ENOENT)
+                        return -errno;
+
+                r = link_find_prioritized(dev, add, dirname, &target);
+                if (r == -ENOENT) {
+                        log_device_debug(dev, "No reference left, removing '%s'", slink);
+                        if (unlink(slink) == 0)
+                                (void) rmdir_parents(slink, "/");
+
+                        break;
+                } else if (r < 0)
+                        return log_device_error_errno(dev, r, "Failed to determine highest priority symlink: %m");
+
+                r = node_symlink(dev, target, slink);
+                if (r < 0) {
+                        (void) unlink(filename);
+                        break;
+                } else if (r == 1)
+                        /* We have replaced already existing symlink, possibly there is some other device trying
+                         * to claim the same symlink. Let's do one more iteration to give us a chance to fix
+                         * the error if other device actually claims the symlink with higher priority. */
+                        continue;
+
+                r = stat(dirname, &st2);
+                if (r < 0 && errno != ENOENT)
+                        return -errno;
+
+                if (stat_inode_unmodified(&st1, &st2))
```

Comment:
Sorry, but that is still wrong (or wronger, because the issue with pointless second stat remains):
We have:
```C
struct stat st1, st2 = {};   // st1 is unitializaed
r = stat(dirname, &st1);  
if (r < 0 && errno != ENOENT) // let's say errno == ENOENT, st1 is still unitialized, `if` is not taken
                        return -errno;
... // nothing touches st1
if (stat_inode_unmodified(&st1, &st2)) // BUG: here the code uses an uninitialized variable
```